lib/matplotlib/axes/_axes.py

```diff
@@ -2208,19 +2208,11 @@ def _convert_dx(dx, x0, xconv, convert):
                 x0 = cbook._safe_first_finite(x0)
             except (TypeError, IndexError, KeyError):
                 pass
-            except StopIteration:
-                # this means we found no finite element, fall back to first
-                # element unconditionally
-                x0 = cbook.safe_first_element(x0)
 
             try:
                 x = cbook._safe_first_finite(xconv)
             except (TypeError, IndexError, KeyError):
```

Comment:
I think there a few other places where StopIteration is try/excepted for _safe_first_element?  Did you check all those?

Reply:
I hadn't, but I have now. I only found these two:

https://github.com/matplotlib/matplotlib/blob/7103779ae295e8a5263e75735082452b3e24e045/lib/matplotlib/dates.py#L1828-L1831

https://github.com/matplotlib/matplotlib/blob/7103779ae295e8a5263e75735082452b3e24e045/lib/matplotlib/units.py#L182-L185

These both predate the change that led to `StopIteration` being thrown for all-nan arrays.  The [commit](https://github.com/matplotlib/matplotlib/commit/af9d6eb9fe59bbc4e5933a0354c5f240fc3a87a0) that introduced the first says it was for zero length objects, and these will still trigger a `StopIteration` when the logic arrives [here](https://github.com/rcomer/matplotlib/blob/f8bc604bd68ef5c20f4e5b0fbdd3c9aabb4d1f7b/lib/matplotlib/cbook.py#L1652).  I'm not sure about the second - should `units.Registry.get_converter` also be able to handle zero length objects?

Reply:
I don't really know - just noted that it happens a few other places, and the more logic we can move into a consistent helper function, the better in my opinion.

Reply:
> the more logic we can move into a consistent helper function, the better in my opinion.

I certainly agree with that.  I wonder if further consolidation should wait for a separate PR though, given recent clarifications about what should and should not be backported 

* my current change fixes a regression from v3.6, so it would be good to get in a patch release
* a change that tidies the code but (hopefully) makes no difference to the user should maybe wait till v3.8